boa/src/builtins/string/mod.rs

```diff
@@ -479,39 +482,90 @@ impl String {
+
+                            match (second, third) {
+                                (Some('$'), _) => {
+                                    // $$
+                                    result.push('$');
+                                }
+                                (Some('&'), _) => {
+                                    // $&
+                                    let matched = caps.get(0).expect("cannot get matched value");
+                                    result.push_str(matched.as_str());
+                                }
+                                (Some('`'), _) => {
+                                    // $`
+                                    let start_of_match = mat.start();
+                                    let slice = &primitive_val[..start_of_match];
+                                    result.push_str(slice);
+                                }
+                                (Some('\''), _) => {
+                                    // $'
+                                    let end_of_match = mat.end();
+                                    let slice = &primitive_val[end_of_match..];
+                                    result.push_str(slice);
+                                }
+                                (_, _) if second_is_digit && third_is_digit => {
+                                    // $nn
+                                    let tens = second.unwrap().to_digit(10).unwrap() as usize;
+                                    let units = third.unwrap().to_digit(10).unwrap() as usize;
+                                    let nn = 10 * tens + units;
+                                    if nn == 0 || nn > m {
+                                        [Some(first), second, chars.next()]
+                                            .iter()
+                                            .flatten()
+                                            .for_each(|ch: &char| result.push(*ch));
+                                    } else {
+                                        let group = match caps.get(nn) {
+                                            Some(text) => text.as_str(),
+                                            None => "",
+                                        };
+                                        result.push_str(group);
+                                        chars.next(); // consume third
+                                    }
+                                }
+                                (_, _) if second_is_digit => {
+                                    // $n
+                                    let n = second.unwrap().to_digit(10).unwrap() as usize;
+                                    if n == 0 || n > m {
+                                        [Some(first), second]
+                                            .iter()
+                                            .flatten()
+                                            .for_each(|ch: &char| result.push(*ch));
```

Comment:
```suggestion
                                (Some(second), _) if second_is_digit => {
                                    // $n
                                    let n = second.to_digit(10).unwrap() as usize;
                                    if n == 0 || n > m {
                                        result.push(first);
                                        result.push(second);
```
Same here we can really simplefy the code and also eliminate a lot of unnesecary calls